File: Framework/Php/framework/framework/collections/smartdictionary.php
<?php

/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */

namespace Smartkernel\Framework\Collections {

    use \SimpleXMLElement;
    use Smartkernel\Framework\SmartString;
    use Smartkernel\Framework\Xml\SmartXml;
    use Smartkernel\Framework\Web\SmartJson;

/*
     * 字典
     */

    class SmartDictionary {
        /*
         * 字典长度（相同的Key算一个）
         * 
         * @param array $array，字典
         * @return int，结果
         * 
         */

        public static function count(array $array) {
            return count($array);
        }

        /*
         * 加入序列
         * 
         * @param array $array，数组
         * @param array $addArray，加入的字典
         * @return array，结果
         * 
         */

        public static function addRange(array $array, array $addArray) {
            foreach ($addArray as $key => $value) {
                $array[$key] = $value;
            }
            return $array;
        }

        /*
         * 排重
         * 
         * @param array $array，字典
         * @return array，结果
         * 
         */

        public static function distinct(array $array) {
            $result = array();
            foreach ($array as $key => $value) {
                if (!self::containsKey($result, $key)) {
                    $result[$key] = $value;
                }
            }
            return $result;
        }

        /*
         * 是否包含
         * 
         * @param array $array，数组
         * @param object $key，键
         * @return bool，结果
         * 
         */

        public static function containsKey(array $array, $key) {
            return array_key_exists($key, $array);
        }

        /*
         * 过滤
         * 
         * @param array $array，数组
         * @param function $where，过滤条件
         * @return array，结果
         * 
         */

        public static function where(array $array, $where) {
            $result = array();
            foreach ($array as $key => $value) {
                if ($where($key, $value) === true) {
                    $result[$key] = $value;
                }
            }
            return $result;
        }

        /*
         * 循环处理
         * 
         * @param array $array，字典
         * @param function $action，函数
         * 
         */

        public static function forEachOne(array $array, $action) {
            foreach ($array as $key => $value) {
                $action($key, $value);
            }
        }

        /*
         * 对象映射为Xml
         * 
         * @param array $item
         *            ，对象
         * @param string $rootName
         *            ，根元素的名称
         * @return string，xml文档
         */

        public static function toXml(array $item, $rootName) {
            $node = new SimpleXMLElement(SmartString::format("<{0} />", $rootName));
            while (list($key, $value) = each($item)) {
                if (is_array($value)) {
                    $subNode = $node->addChild($key);
                    self::toXml($value, $subNode);
                } else {
                    $node->addChild($key, $value);
                }
            }

            return $node->asXml();
        }

        /*
         * 从Xml解析
         * 
         * @param string $xml
         *            ，Xml
         * @return array，结果
         */

        public static function fromXml($xml) {
            return SmartXml::toArray($xml);
        }

        /*
         * 对象映射为Json
         * 
         * @param array $item
         *            ，对象
         * @return string，xml文档
         */

        public static function toJson(array $item) {
            return SmartJson::toJson($item);
        }

        /*
         * 从Json解析
         * 
         * @param string $xml
         *            ，Json
         * @return array，结果
         */

        public static function fromJson($json) {
            return (array) SmartJson::fromJson($json);
        }

    }

}

File: Framework/CSharp/Framework/Framework/Collections/SmartDictionary.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using Smartkernel.Framework.Web;
using System.Collections.Generic;
using System.Text;
using System.Xml.Linq;

namespace Smartkernel.Framework.Collections
{
	/// <summary>
	/// 字典
	/// </summary>
	public class SmartDictionary
	{
		/// <summary>
		/// 对象映射为Xml
		/// </summary>
		/// <param name="item">对象</param>
		/// <param name="rootName">根元素的名称</param>
		/// <returns>xml文档</returns>
		public static string ToXml(Dictionary<string, string> item, string rootName)
		{
			StringBuilder result = new StringBuilder();
			result.Append(string.Format("<{0}>", rootName));
			foreach (var one in item)
			{
				string value = one.Value;
				result.Append(string.Format("<{0}>{1}</{0}>", one.Key, value));
			}
			result.Append(string.Format("</{0}>", rootName));
			return result.ToString();
		}

		/// <summary>
		/// 从Xml解析
		/// </summary>
		/// <param name="xml">Xml</param>
		/// <returns>结果</returns>
		public static Dictionary<string, string> FromXml(string xml)
		{
			Dictionary<string, string> result = new Dictionary<string, string>();

			var nodes = XElement.Parse(xml).Nodes();

			foreach (var node in nodes)
			{
				var nodeXml = node.ToString();
				var element = XElement.Parse(nodeXml);
				result.Add(element.Name.ToString(), element.Value);
			}

			return result;
		}

		/// <summary>
		/// 对象映射为Json
		/// </summary>
		/// <param name="item">对象</param>
		/// <returns>xml文档</returns>
		public static string ToJson(Dictionary<string, string> item)
		{
			return SmartJson.ToJson(item);
		}

		/// <summary>
		/// 从Json解析
		/// </summary>
		/// <param name="json">Json</param>
		/// <returns>结果</returns>
		public static Dictionary<string, string> FromJson(string json)
		{
			return SmartJson.FromJson<Dictionary<string, string>>(json);
		}
	}
}
